core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -443,13 +476,10 @@ class ReplicaManager(val config: KafkaConfig,
         } catch {
           // NOTE: Failed produce requests metric is not incremented for known exceptions
           // it is supposed to indicate un-expected failures of a broker in handling a produce request
-          case e: KafkaStorageException =>
-            fatal("Halting due to unrecoverable I/O error while handling DeleteRecordsRequest: ", e)
-            Runtime.getRuntime.halt(1)
-            (topicPartition, null)
           case e@ (_: UnknownTopicOrPartitionException |
                    _: NotLeaderForPartitionException |
                    _: OffsetOutOfRangeException |
+                   _: KafkaStorageException |
```

Comment:
Good catch! You are right. I have updated the code to call `maybeAddLogFailureEvent` and return KafkaStorageException if it is IOException. I also updated the code not to catch KafkaStorageException.